edx/edx-enterprise
# [ENT-492] Add initial implementation of a generic Consent API.

**Description:** (to quote the upstream ticket): Right now, our consent architecture is fragmented across the `EnterpriseCustomerUser` API and the `EnterpriseCourseEnrollment` API. This PR creates an API that can handle multiple use cases, based on the existing set of data models, in a new consent Djangoapp.

**JIRA:** [ENT-492](https://openedx.atlassian.net/browse/ENT-492)

**Dependencies:** None, but future stories depend on this.

**Merge deadline:** End of this week.

**Testing instructions:**

1. Confirm that Consent API tests cover all important cases.
1. Install this version of `edx-enterprise` in `edx-platform` on a devstack.
1. Register an account on that devstack.
1. Make an `EnterpriseCustomer`.
1. Make an `EnterpriseCustomerUser` based off of your registered account and associate it with the customer from the previous step.
1. Configure a course for the `EnterpriseCustomer` with whatever consent requirements you plan on testing.
1. Enroll in the course, and either provide or do not provide consent for that course, depending on what you plan on testing.
1. Send queries to endpoint `/consent/api/v1/data_sharing_consent` in accordance with the `username` of the account you created, the UUID of the `enterprise_customer` you created, and the `course_id` of the course you enrolled in. Ensure that consent-related responses match with what you planned on testing.
1. If you did not provide consent, send a `POST` request using the `curl` command line tool to the same endpoint (`/consent/api/v1/data_sharing_consent`) and confirm that consent is granted for the course.
1. If you did provide consent, send a `DELETE` request using the `curl` command line tool to the same endpoint (`/consent/api/v1/data_sharing_consent`) and confirm that consent is revoked for that course.

**Merge checklist:**

  **DO NOT** just add dependencies to `requirements/*.txt` files.

**Post merge:**

**Author concerns:** 

1. This is a stepping stone to something more complete; currently we're using models from the `enterprise` app, and we'd like to eventually use models internal to the `consent` app.
1. **The first commit of this application reorganizes the tests to make it easier and more sensible to add tests for new applications.**

**Reviewers**
- @haikuginger 
- @brittneyexline @douglashall
